Rectify rules for attribute inheritance from <g> (#44306)
mdn/content: files/en-us/web/svg/reference/element/g/index.md

This element only includes global attributes.

> Most of the [presentation attributes](https://developer.mozilla.org/en-US/docs/Web/SVG/Reference/Attribute#presentation_attributes) applied to the element are inherited by its children.

Author: The link should be better obviously

Reviewer: MDN links in source are done like this: 

```suggestion
Most of the [presentation attributes](/en-US/docs/Web/SVG/Reference/Attribute#presentation_attributes) applied to the element are inherited by its children.
```

I don't think there is a better link on MDN. Would you be interested in creating one? 
I'm not sure of the right place, but perhaps the simplest way would be to update the link you point to https://developer.mozilla.org/en-US/docs/Web/SVG/Reference/Attribute#presentation_attributes

Under that heading change "All SVG presentation attributes can be used as CSS properties." to something more like "An SVG presentation attribute is an XML attribute that can be used to style an SVG element directly in the markup, but also functions identically as a CSS property. Presentation attributes can be targeted and overridden by external or internal CSS stylesheets."

Then the current note.

Then a new heading structure to separate out the geometric properties.

```
#### Geometry properties (element-specific)

These attributes are Xxxx (explain they aren't inherited)

List of links for these, such as [ry](https://developer.mozilla.org/en-US/docs/Web/SVG/Reference/Attribute/ry) - these explain which are geometry specific.

#### Global presentation attributes

Thes all

[The current list]
```

That's just a suggestion, and if you're not interested keep the current link and we can create an issue to add something later.

Author: I think the information of which rule goes first (between "attributes on g are inherited" and "some attributes are only valid on some elements") should primarily go on the g page, as it's a non-obvious limitation of its main use in practice. I'm not opposed to writing about it also on the properties page as a more theoretical approach.
I'm not bent-on putting my list of all the attributes and the (only) elements they support on _this_ page, it could be better suited on the other.

How about only editing this page in this issue, and then do another to feed the other page and spread the info across the two ?

Reviewer: Sounds fair. For the presentation attributes section, I'm "mostly" interested in providing a definition of what presentation attributes are, because that definition isn't obvious on MDN.

However, it is important to note that this inheritance rule is superseded by SVG's rules for which element supports which attribute.
This means that geometric attributes which are specific to one or several elements will be ignored on a <g> element and not inherited by its children. Such attributes include:
- cx and cy, which are specific to <circle> and <ellipse>
- height, width, x and y, which are specific to <foreignObject>, <image>, <rect>, <svg>, <symbol> and <use>
- r, which is specific to <circle>
- rx and ry, which are specific to <ellipse> and <rect>

> - d, which is specific to <path>

Author: The attributes should be linkified and the elements as well (currently they don't display at all, I don't know what the standard markup is)

Reviewer: Ah, and now I see you are aware of the need to markup. Normally we link as I have done in the other comment. Otherwise on second instance you can do things like `rx` in backticks.

Reviewer: PS I didn't say, but I think your analysis is correct based on my reading of the spec but I am not an expert in SVG by any means.

Other attributes taken by <g>, such as id or class, are not inherited.

> 

Reviewer: Generally we try to avoid wording like "it is important to note". It is important to note (ha ha) that you need to code style (or even better link the elements in markdown), as in markdown they will be treated as elements and that breaks rendering.

Do you think this is accurate still? (and sufficient/better). Note I didn' tlink to the elements on which they are specific, because it doesn't matter for describing the problematic elements. We could add that in the presentation attribute link suggested.

```suggestion
However, SVG-specific element rules supersede general inheritance. 
Geometric attributes specific to certain elements are ignored on a {{SVGElement("g")}} container and **not** inherited by its children. 

These include:
* {{SVGAttr("cx")}}, {{SVGAttr("cy")}}, {{SVGAttr("r")}}: {{SVGElement("circle")}}, {{SVGElement("ellipse")}}
* {{SVGAttr("rx")}}, {{SVGAttr("ry")}}: {{SVGElement("ellipse")}}, {{SVGElement("rect")}}
* {{SVGAttr("d")}}: {{SVGElement("path")}}
* {{SVGAttr("x")}}, {{SVGAttr("y")}}, {{SVGAttr("width")}}, {{SVGAttr("height")}}: {{SVGElement("foreignObject")}}, {{SVGElement("image")}}, {{SVGElement("rect")}}, {{SVGElement("svg")}}, {{SVGElement("symbol")}}, {{SVGElement("use")}}

In addition, some non-geometric attributes valid on {{SVGElement("g")}} such as {{SVGAttr("id")}} and {{SVGAttr("class")}}) are also not inherited by its children.
```

Author: I applied most of this suggestion block. I suggest we continue the contents discussion on the other thread.

This element only includes global attributes.

Most of the [presentation attributes](/en-US/docs/Web/SVG/Reference/Attribute#presentation_attributes) applied to the element are inherited by its children.
However, SVG-specific element rules supersede inheritance.
Geometric attributes specific to certain elements are ignored on a `<g>` container and **not** inherited by its children.

These non-inherited attributes include:
* {{SVGAttr("cx")}}, {{SVGAttr("cy")}}, {{SVGAttr("r")}}: {{SVGElement("circle")}}, {{SVGElement("ellipse")}}
* {{SVGAttr("rx")}}, {{SVGAttr("ry")}}: {{SVGElement("ellipse")}}, {{SVGElement("rect")}}
* {{SVGAttr("d")}}: {{SVGElement("path")}}
* {{SVGAttr("x")}}, {{SVGAttr("y")}}, {{SVGAttr("width")}}, {{SVGAttr("height")}}: {{SVGElement("foreignObject")}}, {{SVGElement("image")}}, {{SVGElement("rect")}}, {{SVGElement("svg")}}, {{SVGElement("symbol")}}, {{SVGElement("use")}}

> In addition, non-presentation attributes valid on `<g>` (such as {{SVGAttr("id")}} or {{SVGAttr("class")}}) are not inherited by its children.

Reviewer: OK, thought about this a bit more. 
1. The issue here is attributes that aren't CSS properties aren't inherited "full stop". Saying that is more clear.
2. `use` is deprecated, so not mentioning it in the list.
3. Not mentioning id and class. They aren't inherited by anything, ever.

If you are also happy with this we can merge.

```suggestion
Most [presentation attributes](/en-US/docs/Web/SVG/Reference/Attribute#presentation_attributes) — those that map to inherited CSS properties, such as {{SVGAttr("fill")}}, {{SVGAttr("stroke")}}, and {{SVGAttr("font-size")}} — are passed down to child elements.
Attributes that are not inherited CSS properties are not passed to children.
This is most notable for geometric attributes, which define the shape or position of a specific element type:

- {{SVGAttr("cx")}}, {{SVGAttr("cy")}}: {{SVGElement("circle")}}, {{SVGElement("ellipse")}}
- {{SVGAttr("r")}}: {{SVGElement("circle")}}
- {{SVGAttr("rx")}}, {{SVGAttr("ry")}}: {{SVGElement("ellipse")}}, {{SVGElement("rect")}}
- {{SVGAttr("d")}}: {{SVGElement("path")}}
- {{SVGAttr("x")}}, {{SVGAttr("y")}}, {{SVGAttr("width")}}, {{SVGAttr("height")}}: {{SVGElement("foreignObject")}}, {{SVGElement("image")}}, {{SVGElement("rect")}}, {{SVGElement("svg")}}, {{SVGElement("symbol")}}
```

Author: 1. The issue is that [r](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/r), [cx](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/cx), [cy](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/cy) and most of the other restricted geometry properties _**are**_ CSS properties as well. So that criterion is incorrect.
2. I don't see [use](https://developer.mozilla.org/en-US/docs/Web/SVG/Reference/Element/use) as deprecated on MDN. What's deprecated is its xlink-scoped attribute.
3. I think id and class are good examples of attributes that g does take, and yet aren't subject to the inheritance meechanism. I think that's efficient to counter the misconception (currently written on the page) that "all attributes are inherited".

Reviewer: > 1. The issue is that [r](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/r), [cx](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/cx), [cy](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/cy) and most of the other restricted geometry properties _**are**_ CSS properties as well. So that criterion is incorrect.

Perhaps I wasn't precise enough (or perhaps I'm just wrong). 

Yes they are CSS properties - the point I'm trying to make is that the are not **inherited** CSS properties: https://www.w3.org/TR/SVG2/geometry.html#CX

So the rule is more "if you are a CSS property that is inherited normally then you will be, if not, you aren't."
The issue with `<g>` is that because it is used as a container for grouping and inheritence, people perhaps expect it to allow everything to inherit, even things that wouldn't otherwise.

> 2. I don't see [use](https://developer.mozilla.org/en-US/docs/Web/SVG/Reference/Element/use) as deprecated on MDN. What's deprecated is its xlink-scoped attribute.

Apologies. You are right. Don't know where my head was.

> 3. I think id and class are good examples of attributes that g does take, and yet aren't subject to the inheritance meechanism. I think that's efficient to counter the misconception (currently written on the page) that "all attributes are inherited".

I guess I don't see them as necessary because they are never inherited by anything, and because we no longer say that all properties are inherited, this assumption is less likely to be made. On the other hand, it does no harm.

Is this more clear?

```suggestion
Most [presentation attributes](https://github.com/en-US/docs/Web/SVG/Reference/Attribute#presentation_attributes) applied to a `<g>` element are inherited by its children — specifically, those that map to inherited CSS properties, such as {{SVGAttr("fill")}}, {{SVGAttr("stroke")}}, and {{SVGAttr("font-size")}}.
Presentation attributes that map to non-inherited CSS properties are not passed to children. This most notably includes geometric attributes such as:

- {{SVGAttr("cx")}}, {{SVGAttr("cy")}}: {{SVGElement("circle")}}, {{SVGElement("ellipse")}}
- {{SVGAttr("r")}}: {{SVGElement("circle")}}
- {{SVGAttr("rx")}}, {{SVGAttr("ry")}}: {{SVGElement("ellipse")}}, {{SVGElement("rect")}}
- {{SVGAttr("d")}}: {{SVGElement("path")}}
- {{SVGAttr("x")}}, {{SVGAttr("y")}}, {{SVGAttr("width")}}, {{SVGAttr("height")}}: {{SVGElement("foreignObject")}}, {{SVGElement("image")}}, {{SVGElement("rect")}}, {{SVGElement("svg")}}, {{SVGElement("symbol")}}, {{SVGElement("use")}}

Non-presentation attributes such as {{SVGAttr("id")}} and {{SVGAttr("class")}} are also not inherited becaues they are not part of the CSS cascade.
```

The main thing I am trying to avoid is "However, SVG-specific element rules supersede inheritance.". It is imprecise.

Author: You're presenting things so that the inheritance is shown as a superpower not of the g element, but of the CSS properties themselves. (It does not seem to be wrong, but.)

In that version you are removing the information that the listed properties (regardless of CSS) are specific to one or more SVG elements. And so then, it's not clear why elements are listed after each property in the list.

Furthermore, I don't find that version of the rule to be more intuitive, nor easy to check for a particular attribute/property.

Finally, the explanation in the specification I linked is closer to "mine" than "yours" :
> Most SVG presentation attributes may be specified on any element in the SVG namespace where there is not a name clash with an existing attribute. However, the geometry properties only have equivalent presentation attributes on designated elements.

It specifically presents the rule as something specific *to the element*, not *to the CSS property* and does not even mention the inheritability of the CSS property in general (= when not set through an attribute).

Reviewer: Thanks. You have convinced me. Still don't like that wording of "However, SVG-specific element rules supersede inheritance.".  So if we could just tidy that: https://github.com/mdn/content/pull/44306/changes#r3371284933

Author: I did that and also rephrased the last line a bit.

This element only includes global attributes.

Most of the [presentation attributes](/en-US/docs/Web/SVG/Reference/Attribute#presentation_attributes) applied to the element are inherited by its children.
However, SVG-specific element rules supersede inheritance.

> Geometric attributes specific to certain elements are ignored on a `<g>` container and **not** inherited by its children.

Reviewer: ```suggestion
However, geometric attributes are only valid as presentation attributes on their designated elements — they have no effect on a `<g>` element and are not passed to its children.
```
